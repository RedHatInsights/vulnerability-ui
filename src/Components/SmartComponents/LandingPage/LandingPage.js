import React, { Fragment } from 'react';
import { FormattedMessage } from 'react-intl';
import { Popover } from '@patternfly/react-core';
import { OutlinedQuestionCircleIcon, ExternalLinkAltIcon } from '@patternfly/react-icons';
import messages from '../../../Messages';
import { Main } from '@redhat-cloud-services/frontend-components/Main';
import Header from '../../PresentationalComponents/Header/Header';
import CVEs from '../CVEs/CVEs';
import Dashbar from '../Dashbar/Dashbar';

const LandingPage = () => {
    // eslint-disable-next-line max-len
    const PRODUCT_DOC = 'https://access.redhat.com/documentation/en-us/red_hat_insights/2022/html/assessing_and_monitoring_security_vulnerabilities_on_rhel_systems/index';

    const title = (
        <Popover
            enableFlip
            position="right"
            headerContent={<FormattedMessage {...messages.ovalPopoverHeader} />}
            bodyContent={<FormattedMessage {...messages.ovalPopoverBody} />}
            footerContent={<a href={PRODUCT_DOC} target="__blank" rel="noopener noreferrer">
                <FormattedMessage {...messages.learnMore} /> <ExternalLinkAltIcon />
            </a>}
        >
            <span>
                <FormattedMessage {...messages.cvesHeader} />
                <OutlinedQuestionCircleIcon
                    color="var(--pf-global--secondary-color--100)"
                    className="pf-u-ml-sm pointer cves-header-questionmark"
                    style={{ verticalAlign: '0' }}
                />
            </span>
        </Popover>
    );

    return (
        <Fragment>
            <Header title={title} showBreadcrumb={false} />
            <Dashbar />
            <Main>
                <CVEs />
            </Main>
        </Fragment>
    );
};

export default LandingPage;
