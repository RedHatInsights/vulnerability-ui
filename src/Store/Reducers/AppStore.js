import * as ActionTypes from '../ActionTypes';

// Initial State
export const initialState = {
    popup: {}
};

// Reducer
export const AppStore = (state = initialState, action) => {
    switch (action.type) {
        case ActionTypes.TOGGLE_POPUP: {
            const { page, state } = action.payload;
            return {
                ...state,
                popup: {
                    ...state.popup,
                    [page]: state
                }
            };
        }

        default:
            return state;
    }
};
