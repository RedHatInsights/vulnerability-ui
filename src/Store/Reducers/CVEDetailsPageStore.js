import Immutable from 'seamless-immutable';
import * as ActionTypes from '../ActionTypes';

// Initial State
export const initialState = Immutable({
    affectedSystemsByCVE: {
        isLoading: true,
        openedRows: [],
        payload: {
            meta: {
                total_items: 0,
                page: 1,
                page_size: 20
            }
        }
    },
    parameters: {
        page: 1,
        page_size: 20,
        selectedHosts: [],
        status_id: undefined,
        sort: '-last_upload',
        show_advisories: true
    },
    cveDetails: {
        isLoading: true,
        payload: {},
        error: false
    }
});

// Reducer
export const CVEDetailsPageStore = (state = initialState, action) => {
    let newState;
    switch (action.type) {
        case ActionTypes.FETCH_CVE_DETAILS + '_REJECTED':
            newState = state.setIn(['cveDetails', 'error'], true);
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;
        case ActionTypes.FETCH_CVE_DETAILS + '_PENDING':
            newState = state.setIn(['cveDetails', 'isLoading'], true);
            return newState;

        case ActionTypes.FETCH_CVE_DETAILS + '_FULFILLED':
            newState = state.setIn(['cveDetails', 'payload'], action.payload);
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;

        case ActionTypes.FETCH_AFFECTED_SYSTEMS_BY_CVE + '_PENDING':
            newState = state.setIn(['affectedSystemsByCVE', 'isLoading'], true);
            return newState;

        case ActionTypes.FETCH_AFFECTED_SYSTEMS_BY_CVE + '_FULFILLED':
            newState = state.setIn(['affectedSystemsByCVE', 'payload'], action.payload);
            newState = newState.setIn(['affectedSystemsByCVE', 'isLoading'], false);
            return newState;
        case ActionTypes.CHANGE_EXPOSED_SYSTEMS_PARAMETERS:
            newState = state.setIn(['parameters'], {
                ...state.parameters,
                ...action.payload,
                page_size: action.payload.page_size || state.parameters.page_size
            });
            return newState;
        case ActionTypes.SET_GLOBAL_FILTER:
            newState = state.setIn(['parameters'], {
                ...state.parameters,
                ...action.payload
            });
            return newState;
        case ActionTypes.CLEAR_CVE_STORE:
            return initialState;
        default:
            return state;
    }
};

