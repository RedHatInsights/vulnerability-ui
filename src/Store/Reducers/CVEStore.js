import Immutable from 'seamless-immutable';
import * as ActionTypes from '../../Store/ActionTypes';

// Initial State
export const initialState = Immutable({
    affectedSystemsByCVE: {
        isLoading: true,
        payload: [],
        openedRows: []
    },
    cveDetails: {
        isLoading: true,
        payload: {},
        error: false
    }
});

// Reducer
export const CVEReducer = (state = initialState, action) => {
    let newState;
    let openedRows;
    switch (action.type) {
        case ActionTypes.HANDLE_CVE_DETAILS_ERROR:
            newState = state.setIn(['cveDetails', 'error'], action.payload.error);
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;
        case ActionTypes.FETCH_CVE_DETAILS + '_PENDING':
            newState = state.setIn(['cveDetails', 'isLoading'], true);
            return newState;

        case ActionTypes.FETCH_CVE_DETAILS + '_FULFILLED':
            newState = state.setIn(['cveDetails', 'payload'], action.payload);
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;

        case ActionTypes.FETCH_AFFECTED_SYSTEMS_BY_CVE + '_PENDING':
            newState = state.setIn(['affectedSystemsByCVE', 'isLoading'], true);
            return newState;

        case ActionTypes.FETCH_AFFECTED_SYSTEMS_BY_CVE + '_FULFILLED':
            newState = state.setIn(['affectedSystemsByCVE', 'payload'], action.payload);
            newState = newState.setIn(['affectedSystemsByCVE', 'isLoading'], false);
            return newState;
        case ActionTypes.EXPAND_ROW:
            openedRows = [...state.affectedSystemsByCVE.openedRows];

            if (openedRows.includes(action.payload.id)) {
                openedRows.splice(openedRows.indexOf(action.payload.id), 1);
            } else {
                openedRows = [action.payload.id, ...openedRows];
            }

            newState = state.setIn(['affectedSystemsByCVE', 'openedRows'], openedRows);
            return newState;
        default:
            return state;
    }
};
