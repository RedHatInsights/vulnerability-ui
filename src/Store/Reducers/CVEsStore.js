import * as ActionTypes from '../ActionTypes';
import { applyGlobalFilter, isTimestampValid } from './reducersHelper';

// Initial State
export const initialState = {
    cveList: {
        isLoading: true,
        payload: [],
        timestamp: new Date()
    },
    parameters: {
        sort: '-public_date',
        affecting: 'true,false'
    },
    selectedCves: [],
    expandedRows: [],
    isAllExpanded: false
};

// Reducer
export const CVEsStore = (state = initialState, action) => {
    let newState = { ...state };

    switch (action.type) {
        case ActionTypes.FETCH_CVE_LIST + '_PENDING':
            newState.cveList = {
                timestamp: action.meta,
                isLoading: true
            };
            return newState;

        case ActionTypes.FETCH_CVE_LIST + '_REJECTED':
            newState.error = true;
            newState.cveList = {
                isLoading: false,
                payload: { errors: action.payload }
            };
            return newState;

        case ActionTypes.FETCH_CVE_LIST + '_FULFILLED':
            if (isTimestampValid(state.cveList.timestamp, action.meta)) {
                newState = {
                    ...state,
                    cveList: {
                        ...state.cveList,
                        payload: action.payload,
                        isLoading: false
                    },
                    ...state.isAllExpanded && { expandedRows: action.payload.data.map(({ id }) => id) }
                };

                return newState;
            }

            return state;

        case ActionTypes.CHANGE_CVE_LIST_PARAMETERS + '_FULFILLED':
            newState.parameters = { ...newState.parameters, ...action.payload };
            return newState;

        case ActionTypes.SELECT_CVE: {
            let selectedCves = newState.selectedCves.slice();
            if (Array.isArray(action.payload)) {
                selectedCves = action.payload;
            } else {
                (selectedCves.includes(action.payload) && selectedCves.splice(selectedCves.indexOf(action.payload), 1)) ||
                    selectedCves.push(action.payload);
            }

            return { ...newState, selectedCves  };
        }

        case ActionTypes.EXPAND_CVE: {
            if (Array.isArray(action.payload)) {
                const expandedRows = action.payload;
                const isAllExpanded = action.payload.length === 0 ? false : true;
                return { ...newState,  expandedRows, isAllExpanded };
            }

            const cveName = newState.cveList.payload.data[action.payload / 2].id;
            const expandedRows = newState.expandedRows.slice();
            expandedRows.includes(cveName) && expandedRows.splice(expandedRows.indexOf(cveName), 1)
            || expandedRows.push(cveName);

            return { ...newState, expandedRows };
        }

        case ActionTypes.CLEAR_CVES_STORE:
            return initialState;

        case ActionTypes.SET_GLOBAL_FILTER:
            newState.parameters = {
                ...applyGlobalFilter(newState.parameters, action.payload)
            };
            return newState;

        default:
            return state;
    }
};
