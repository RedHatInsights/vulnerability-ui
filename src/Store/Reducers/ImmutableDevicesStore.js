import Immutable from 'seamless-immutable';
import * as ActionTypes from '../ActionTypes';
import { applyGlobalFilter } from './reducersHelper';
import { DEFAULT_PAGE_SIZE } from '../../Helpers/constants';

export const initialState = Immutable({
    parameters: {
        page: 1,
        page_size: DEFAULT_PAGE_SIZE,
        sort: '-updated',
        //makes sure only edge devices are shown
        host_type: 'edge'
    },
    devices: {
        isLoading: true,
        payload: {},
        error: {
            hasError: false
        }
    }
});

export const ImmutableDevicesStore = (state = initialState, action) => {
    let newState;

    switch (action.type) {
        case ActionTypes.FETCH_CVE_DEVICES + '_REJECTED':
            newState = state.setIn(['cveDetails', 'error'], { hasError: true, errorCode: action.payload.status });
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;
        case ActionTypes.FETCH_CVE_DEVICES + '_PENDING':
            newState = state.setIn(['cveDetails', 'isLoading'], true);
            return newState;

        case ActionTypes.FETCH_CVE_DEVICES + '_FULFILLED':
            newState = state.setIn(['cveDetails', 'payload'], action.payload);
            newState = newState.setIn(['cveDetails', 'isLoading'], false);
            return newState;

        case ActionTypes.CHANGE_EXPOSED_DEVICES_PARAMETERS: {
            const { reset, ...params } = action.payload;

            newState = state.setIn(['parameters'], {
                ...!reset ? state.parameters : {},
                ...params,
                page_size: action.payload.page_size || state.parameters.page_size
            });
            return newState;
        }

        case ActionTypes.SET_GLOBAL_FILTER:
            newState = state.setIn(['parameters'], {
                ...applyGlobalFilter(state.parameters, action.payload)
            });
            return newState;
        case ActionTypes.CLEAR_DEVICES_STORE:
            return initialState;
        default:
            return state;
    }
};

