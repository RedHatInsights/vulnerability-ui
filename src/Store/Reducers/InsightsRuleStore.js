import * as ActionTypes from '../ActionTypes';

const initialState = {
    error: false,
    isLoading: true,
    systemRules: []
};

export const InsightsRuleStore = (state = initialState, action) => {
    let newState = { ...state };
    switch (action.type) {
        case ActionTypes.FETCH_INSIGHTS_SYSTEM_RULES + '_PENDING':
            return newState;
        case ActionTypes.FETCH_INSIGHTS_SYSTEM_RULES + '_REJECTED':
            newState.isLoading = false;
            newState.error = true;
            return newState;
        case ActionTypes.FETCH_INSIGHTS_SYSTEM_RULES + '_FULFILLED':
            newState.isLoading = false;
            newState.systemRules = action.payload;
            return newState;
        default:
            return state;
    }
};
