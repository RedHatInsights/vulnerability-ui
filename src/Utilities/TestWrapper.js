import propTypes from 'prop-types';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

const TestWrapper = ({ children, store }) => {
    return (
        <IntlProvider locale="en">
            { store ?
                <Provider store={store}>
                    <Router>
                        {children}
                    </Router>
                </Provider> : children
            }
        </IntlProvider>
    );
};

TestWrapper.propTypes = {
    children: propTypes.node,
    store: propTypes.object
};

export default TestWrapper;
