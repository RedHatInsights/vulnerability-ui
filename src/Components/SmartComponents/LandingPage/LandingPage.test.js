import LandingPage from './LandingPage';
import TestWrapper from '../../../Utilities/TestWrapper';
import { render, screen } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { cveState } from '../mockData.fixtures';
import { initialState } from '../../../Store/Reducers/CVEsStore';
import { useSelector } from 'react-redux';
import '@testing-library/jest-dom';

jest.mock('../../../Helpers/Hooks', () => ({
    ...jest.requireActual('../../../Helpers/Hooks'),
    useRbac: () => [[true, true, true, true], false],
    useHybridSystemFilterFlag: jest.fn(() => false)
}));

jest.mock('../../../Store/Actions/Actions', () => ({
    ...jest.requireActual('../../../Store/Actions/Actions'),
    fetchCveListByAccount: () => ({
        type: 'FETCH_CVE_LIST',
        payload: new Promise(() => {})
    })
}));

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn()
}));

jest.mock('@unleash/proxy-client-react', () => ( {
    ...jest.requireActual('@unleash/proxy-client-react'),
    useFlag: () => true, 
    useFlagsStatus: () => ({ flagsReady: true })
}));

jest.mock('../../../Helpers/APIHelper');

useSelector.mockImplementation(callback => {
    return callback({ CVEsStore: cveState });
});

const mockStore = configureStore([]);
const store = mockStore(initialState);


describe('CVEs: ', () => {
    it('Should match the snapshot', () => {
        const { asFragment } = render(
            <TestWrapper store={ store }>
                <LandingPage />
            </TestWrapper>
        );
        expect(asFragment()).toMatchSnapshot();
    });

    it('Should render Header with props showBreadcrumb = false', () => {
        render(
            <TestWrapper store={ store }>
                <LandingPage />
            </TestWrapper>
        );

        expect(screen.getByRole('heading', { name: 'CVEs' })).toBeVisible();
        expect(screen.queryAllByTestId('PfBreadcrumb')).toHaveLength(0);
    });

    it('Should render VulnerabilitiesCves', () => {
        render(
            <TestWrapper store={ store }>
                <LandingPage />
            </TestWrapper>
        );

        expect(screen.getByRole('grid', { name: 'Vulnerability CVE table' })).toBeVisible();
    });
});
