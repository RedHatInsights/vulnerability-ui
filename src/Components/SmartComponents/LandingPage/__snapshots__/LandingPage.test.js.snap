// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CVEs:  Should match the snapshot 1`] = `
<Fragment>
  <Header
    showBreadcrumb={false}
    title="CVEs"
  />
  <Connect(Main)>
    <CVEs />
    <PopUp
      body="Tag and workload filtration is temporarily unavailable on this page. We're working to resolve this soon."
      page="landing"
      title="Tag and workload filtration temporarily unavailable"
    />
  </Connect(Main)>
</Fragment>
`;
